lib/ABI/ApplyRegisterStateDeductions.cpp

```diff
@@ -0,0 +1,292 @@
+/// \file ApplyRegisterStateDeductions.cpp
+/// \brief
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/ArrayRef.h"
+
+#include "revng/ABI/ApplyRegisterStateDeductions.h"
+#include "revng/ABI/CallingConventionTrait.h"
+#include "revng/Model/Types.h"
+#include "revng/Support/EnumSwitch.h"
+
+/// Finds last relevant (`Yes` or `Dead`) register out of the provided list.
+///
+/// Returns `std::numeric_limits<size_t>::max()` when no registers from
+/// the list were used.
+template<typename AccessorType>
+size_t findLastUsedIndex(llvm::ArrayRef<model::Register::Values> Registers,
+                         const RegisterStateMap &State,
+                         const AccessorType &Accessor) {
+  for (size_t Index = Registers.size(); Index != 0; --Index)
+    if (model::RegisterState::isYesOrDead(Accessor(State.at(Registers[Index]))))
+      return Index;
+
+  return std::numeric_limits<size_t>::max();
+}
+
+/// Limits the passed range to only the elements before the last one used
+/// \sa findLastUsedIndex()
+template<typename AccessorType>
+llvm::ArrayRef<model::Register::Values>
+getRegisterSubrange(llvm::ArrayRef<model::Register::Values> Registers,
+                    const RegisterStateMap &State,
+                    const AccessorType &Accessor) {
```

Comment:
I feel like this function's name should contain "last used" or something like that.

Reply:
Or accept the index, yeah, makes sense.

Reply:
Gone (https://github.com/revng/revng/commit/f06daab3a2ce9a6ef1a4e357579c4cf7c16a57b8).